zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResult.java

```diff
@@ -87,36 +89,51 @@ private String getData(String msg) {
     if (msg == null) {
       return null;
     }
-
-    Type[] types = Type.values();
-    for (Type t : types) {
-      String magic = "%" + t.name().toLowerCase();
-      if (msg.startsWith(magic + " ") || msg.startsWith(magic + "\n")) {
-        int magicLength = magic.length() + 1;
-        if (msg.length() > magicLength) {
-          return msg.substring(magicLength);
-        } else {
-          return "";
-        }
-      }
+    Type[] types = type.values();
+    TreeMap<Integer, Type> typesLastIndexInMsg = buildIndexMap(msg);
+    if (typesLastIndexInMsg.size() == 0) {
+      return msg;
+    } else {
+      Map.Entry<Integer, Type> lastType = typesLastIndexInMsg.lastEntry();
+      int magicLength = lastType.getValue().name().length() + 1; //add 1 for the %
+      int subStringPos = magicLength + lastType.getKey() + 1; // 1 for the last \w after magic
```

Comment:
\w? do you mean newline or regex:\w?
if latter, what if someone types "%" - ie. no character following %